### PR TITLE
Eliminate D2item

### PR DESCRIPTION
**Pull request by bhollis**
Destiny 2 is definitely the default for DIM, and it's where all our development is going. Destiny 3 is unlikely to ever happen. This PR removes the "D2Item" type and folds all its properties back into "DimItem" - they were all optional anyway. This means we no longer have to do `item.isDestiny2()` just to get access to some properties. "D1Item" remains because it has some weird stuff, but unless you care about that weird stuff you'll never have to touch it.

Along the way I realized I could finally typecheck the big property blob in the item factory methods, and that helped out a lot.

The "Year", "Season", and "Event" properties have been totally removed and are just calculated on-demand.

### CODE REVIEW

**Comment by nev-r** [comment]
oh yay

File: src/app/inventory/item-types.ts [hunk 1]
```diff
@@ -161,14 +160,42 @@ export interface DimItem {
...
 
   /** "Touch" the item to mark it as having been manually moved. */
   updateManualMoveTimestamp(): void;
 
   /** Check if this item is from D1. Inside an if statement, this item will be narrowed to type D1Item. */
   isDestiny1(): this is D1Item;
-  /** Check if this item is from D2. Inside an if statement, this item will be narrowed to type D2Item. */
-  isDestiny2(): this is D2Item;
+  /** Check if this item is from D2. */
+  isDestiny2(): boolean;
```
Comment (kyleshay) [inline src/app/inventory/item-types.ts:198]: is it still being used in some places?

**Review by kyleshay — APPROVED** [review]
awesome!

File: src/app/inventory/item-types.ts (hunk 1 above)
Comment (bhollis) [reply src/app/inventory/item-types.ts:198]: Yup - there are places where we still want to do something only for D2, and aren't using it for type info.

File: src/app/inventory/spreadsheets.ts [hunk 2]
```diff
@@ -272,11 +276,11 @@ function equippable(item: DimItem) {
 }
 
 export function source(item: DimItem) {
-  if (item.isDestiny2()) {
+  if (item.source) {
     return (
       sourceKeys.find(
         (src) =>
-          D2Sources[src].sourceHashes.includes(item.source) ||
+          (item.source && D2Sources[src].sourceHashes.includes(item.source)) ||
```
Comment (nev-r) [inline src/app/inventory/spreadsheets.ts:283]: dammit typescript

File: src/app/inventory/store/d2-item-factory.ts [hunk 3]
```diff
@@ -54,16 +51,16 @@ const collectiblesByItemHash = _.once(
...
   // Mark that this item has been moved manually
-  updateManualMoveTimestamp(this: D2Item) {
+  updateManualMoveTimestamp(this: DimItem) {
     this.lastManuallyMoved = Date.now();
     if (this.id !== '0') {
       _moveTouchTimestamps.set(this.id, this.lastManuallyMoved);
     }
   },
-  isDestiny1(this: D2Item) {
+  isDestiny1(this: DimItem) {
```
Comment (nev-r) [inline src/app/inventory/store/d2-item-factory.ts:60]: are these 2 `this`es doing anything?

File: src/app/item-popup/EnergyMeter.tsx [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
 import './EnergyMeter.scss';
 
-export default function EnergyMeter({ defs, item }: { defs: D2ManifestDefinitions; item: D2Item }) {
+export default function EnergyMeter({
+  defs,
+  item,
+}: {
+  defs: D2ManifestDefinitions;
+  item: DimItem;
+}) {
```
Comment (nev-r) [inline src/app/item-popup/EnergyMeter.tsx:15]: can't believe this was just 1 character away from wrapping

**Review by nev-r — APPROVED** [review]

File: src/app/inventory/store/d2-item-factory.ts (hunk 3 above)
Comment (bhollis) [reply src/app/inventory/store/d2-item-factory.ts:60]: Nope

**Comment by ryan-rushton** [comment]
Just got around to looking at this and it's a very welcome change

**Comment by wickles** [comment]
2020-09-14 19_22_17-ShareX-chrome

**Comment by nev-r** [comment]
@wickles should be good next time there's an update dot for you. thank you for the report!

**Pull request merged**